controllers/vaultstaticsecret_controller.go

```diff
@@ -218,11 +245,242 @@ func (r *VaultStaticSecretReconciler) handleDeletion(ctx context.Context, o clie
+	// to close this watcher and requeue the resource
+	r.SourceCh <- event.GenericEvent{
+		Object: &secretsv1beta1.VaultStaticSecret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: o.Namespace,
+				Name:      o.Name,
+			},
+		},
+	}
+
+	return nil
+}
+
+// eventMsg is used to extract the relevant fields from an event message sent
+// from Vault
+type eventMsg struct {
+	Data struct {
+		Event struct {
+			Metadata struct {
+				Path     string `json:"path"`
+				Modified string `json:"modified"`
+			} `json:"metadata"`
+		} `json:"event"`
+		Namespace string `json:"namespace"`
+	} `json:"data"`
+}
+
+func (r *VaultStaticSecretReconciler) streamStaticSecretEvents(ctx context.Context, o *secretsv1beta1.VaultStaticSecret, wsClient *vault.WebsocketClient) error {
+	logger := log.FromContext(ctx)
+	conn, err := wsClient.Connect(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to connect to vault websocket: %w", err)
+	}
+	defer conn.Close(websocket.StatusNormalClosure, "closing event watcher")
+
+	// We made it past the initial websocket connection, so emit a "good" event
+	// status
+	r.Recorder.Eventf(o, corev1.EventTypeNormal, consts.ReasonEventWatcherStarted, "Started watching events")
+
+	for {
+		select {
+		case <-ctx.Done():
+			logger.V(consts.LogLevelDebug).Info("Context done, closing websocket",
+				"namespace", o.Namespace, "name", o.Name)
+			return nil
+		default:
+			msgType, message, err := conn.Read(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to read from websocket: %w, message: %q",
+					err, string(message))
+			}
+			messageMap := eventMsg{}
+			err = json.Unmarshal(message, &messageMap)
+			if err != nil {
+				return fmt.Errorf("failed to unmarshal event message: %w", err)
+			}
+			logger.V(consts.LogLevelTrace).Info("Received message", "message type", msgType, "message", messageMap)
+			modified := messageMap.Data.Event.Metadata.Modified
+
+			if modified == "true" {
+				namespace := strings.Trim(messageMap.Data.Namespace, "/")
+				path := messageMap.Data.Event.Metadata.Path
+				specPath := strings.Join([]string{o.Spec.Mount, o.Spec.Path}, "/")
+
+				if o.Spec.Type == consts.KVSecretTypeV2 {
+					specPath = strings.Join([]string{o.Spec.Mount, "data", o.Spec.Path}, "/")
+				}
+				logger.V(consts.LogLevelTrace).Info("modified Event received from Vault",
+					"namespace", namespace, "path", path, "spec.namespace", o.Spec.Namespace,
+					"spec path", specPath)
+				if namespace == o.Spec.Namespace && path == specPath {
+					logger.V(consts.LogLevelDebug).Info("Event matches, sending requeue", "namespace", namespace, "path", path)
+					r.SourceCh <- event.GenericEvent{
+						Object: &secretsv1beta1.VaultStaticSecret{
+							ObjectMeta: metav1.ObjectMeta{
+								Namespace: o.Namespace,
+								Name:      o.Name,
+							},
+						},
+					}
+				}
+			} else {
+				// This is an event we're not interested in, ignore it and
+				// carry on.
+				continue
```

Comment:
Could it be beneficial to add a trace level log call here?

Reply:
Sure, added one in 2b28b51